### PR TITLE
MP-2615 rover rewards

### PR DESCRIPTION
**Pull request by piobab**
Add support to track rover account ids in deposit/withdraw

Note: All migration for red bank will go in separate PR.

### CODE REVIEW

File: contracts/incentives/src/state.rs [hunk 1]
```diff
@@ -32,12 +32,12 @@ pub const INCENTIVE_STATES: Map<(&str, &str), IncentiveState> = Map::new("incent
 pub const EMISSIONS: Map<(&str, &str, u64), Uint128> = Map::new("emissions");
 
 /// A map containing the incentive index for a given user, collateral denom and incentive denom.
-/// The key is (user address, collateral denom, incentive denom).
-pub const USER_ASSET_INDICES: Map<(&Addr, &str, &str), Decimal> = Map::new("indices");
+/// The key is (user address or account id, collateral denom, incentive denom).
+pub const USER_ASSET_INDICES: Map<(&str, &str, &str), Decimal> = Map::new("indices");
```
Comment (piobab) [inline contracts/incentives/src/state.rs:36]: Hmm maybe better to keep (&Addr, &str) even duplicating `Addr` for rover accounts. It can be consistent with redbank.

Comment (piobab) [reply contracts/incentives/src/state.rs:36]: Added in another commit. Will revert in case not accepted by others

File: contracts/incentives/src/state.rs [hunk 2]
```diff
@@ -32,12 +32,13 @@ pub const INCENTIVE_STATES: Map<(&str, &str), IncentiveState> = Map::new("incent
 pub const EMISSIONS: Map<(&str, &str, u64), Uint128> = Map::new("emissions");
 
 /// A map containing the incentive index for a given user, collateral denom and incentive denom.
-/// The key is (user address, collateral denom, incentive denom).
-pub const USER_ASSET_INDICES: Map<(&Addr, &str, &str), Decimal> = Map::new("indices");
+/// The key is (user address or account id, collateral denom, incentive denom).
```
Comment (grod220) [inline contracts/incentives/src/state.rs:35]: isn't it `user address with optional account id`

File: contracts/incentives/tests/test_balance_change.rs [hunk 3]
```diff
@@ -278,9 +284,16 @@ fn with_zero_previous_balance_and_asset_with_zero_index_accumulates_rewards() {
...
-            query_user_unclaimed_rewards(deps.as_ref(), env, "user".to_string(), None, None, None)
-                .unwrap();
+        let rewards_query = query_user_unclaimed_rewards(
+            deps.as_ref(),
+            env,
+            "user".to_string(),
+            None,
+            None,
+            None,
+            None,
```
Comment (grod220) [inline contracts/incentives/tests/test_balance_change.rs:294]: Yeesh, so many none's. Wish rust had named arguments lol.

File: contracts/incentives/tests/test_balance_change.rs
Comment (grod220) [inline contracts/incentives/tests/test_balance_change.rs:1]: Do we need an account_id test here? They all seem none.

File: contracts/incentives/tests/test_claim_rewards.rs
Comment (grod220) [inline contracts/incentives/tests/test_claim_rewards.rs:1]: An account_id test here too?

File: contracts/red-bank/src/collateral.rs [hunk 4]
```diff
@@ -21,18 +21,17 @@ pub fn update_asset_collateral_status(
...
+        .ok_or_else(|| ContractError::UserNoCollateralBalance {
+            user: user.into(),
+            denom: denom.clone(),
         })?;
 
     let previously_enabled = collateral.enabled;
 
     collateral.enabled = enable;
-    COLLATERALS.save(deps.storage, (user.address(), &denom), &collateral)?;
+    COLLATERALS.save(deps.storage, (user.address(), "", &denom), &collateral)?;
```
Comment (grod220) [inline contracts/red-bank/src/collateral.rs:34]: Wouldn't this still allow them to use the ones that are associated with account ids?

File: contracts/red-bank/src/health.rs [hunk 5]
```diff
@@ -117,7 +117,7 @@ pub fn get_user_positions_map(
 
     // Find all denoms that the user has a collateral or debt position in
     let collateral_denoms = COLLATERALS
-        .prefix(user_addr)
+        .prefix((user_addr, ""))
```
Comment (grod220) [inline contracts/red-bank/src/health.rs:120]: We shouldn't assume none here right?

**Review by grod220 — COMMENTED** [review]
There are usages in the contracts where account ID is assumed to be `""`. I wonder if that poses safety issues. Shouldn't all of those function signatures accept `account_id` now?

File: contracts/red-bank/src/collateral.rs (hunk 4 above)
Comment (piobab) [reply contracts/red-bank/src/collateral.rs:34]: We don't support enabling / disabling collateral for rover accounts so I think no. By default collateral for rover accounts are enabled.

File: contracts/red-bank/src/health.rs (hunk 5 above)
Comment (piobab) [reply contracts/red-bank/src/health.rs:120]: hmm maybe for safety I will add account_id: Option<String>. We handle health in different way for redbank and rover so basically this function is used by redbank users.

File: contracts/incentives/tests/test_claim_rewards.rs
Comment (piobab) [reply contracts/incentives/tests/test_claim_rewards.rs:1]: see `rewards_claim_for_credit_account`.

File: contracts/incentives/tests/test_balance_change.rs (hunk 3 above)
Comment (piobab) [reply contracts/incentives/tests/test_balance_change.rs:294]: Yeap, I wish too.

**Review by grod220 — APPROVED** [review]

**Review by brimigs — APPROVED** [review]
